.test-infra/metrics/sync/github/sync.py

```diff
@@ -0,0 +1,398 @@
+  return list(map(lambda x: x["node"]["requestedReviewer"]["login"], reviewEdges))
+
+
+def extractMentions(pr):
+  body = pr["body"]
+  commentEdges = pr["comments"]["edges"]
+  reviewEdges = pr["reviews"]["edges"]
+
+  bodyMentions = ghutilities.findMentions(body)
+  commentMentionsLists = map(lambda x: ghutilities.findMentions(x["node"]["body"]),
+                             commentEdges)
+  reviewMentionsLists = map(lambda x: ghutilities.findMentions(x["node"]["body"]),
+                            commentEdges)
+  commentMentions = [item for sublist in commentMentionsLists for item in sublist]
+  reviewMentions = [item for sublist in reviewMentionsLists for item in sublist]
+
+  mentionsSet = set(bodyMentions) | set(commentMentions) | set(reviewMentions)
+  return list(mentionsSet)
+
+
+def extractFirstNAActivity(pr):
+  '''Returns timestamp and login of author on first activity on pull request done by non-author.'''
+  author = pr["author"]["login"]
+  commentEdges = None
+  commentEdges = [edge for edge in pr["comments"]["edges"] if edge["node"]["author"]["login"] != author]
+  reviewEdges = [edge for edge in pr["reviews"]["edges"] if edge["node"]["author"]["login"] != author]
+  merged = pr["merged"]
+  mergedAt = pr["mergedAt"]
+  mergedBy = None if not merged else pr["mergedBy"]["login"]
+  commentTimestamps = list(map(lambda x: (x["node"]["createdAt"], x["node"]["author"]["login"]),
+                               commentEdges))
+  reviewTimestamps = list(map(lambda x: (x["node"]["createdAt"], x["node"]["author"]["login"]),
+                              reviewEdges))
+  allTimestamps = commentTimestamps + reviewTimestamps
+  if merged:
+    allTimestamps.append((mergedAt, mergedBy))
+  return (None, None) if not allTimestamps else min(allTimestamps, key=lambda t: t[0])
+
+
+def extractBeamReviewers(pr):
+  '''Extract logins of users defined by Beam as reviewers.'''
+  author = pr['author']['login']
+
+  # All the direct GitHub indicators of reviewers
+  reviewers = []
+  for r in pr['assignees']['edges']:
+    reviewers.append(r['node']['login'])
+  for r in pr['reviewRequests']['edges']:
+    reviewers.append(r['node']['requestedReviewer']['login'])
+
+  # GitHub users that have performed reviews.
+  for r in pr['reviews']['edges']:
+    reviewers.append(r['node']['author']['login'])
+
+  ## Hi, @r1, can you .. look?
+  beamReviewerRegex = r'(Hi.*look?)'
+  ## R= @r1 @r2 @R3
+  g3ReviewerRegex = r'(?:^|\W)[Rr]\s*[=:.]((?:[\s,;.]*-?@\w+)+)'
+  usernameRegex = r'(-?)(@\w+)'
+  for m in [pr['body']] + [c['node']['body'] for c in pr['comments']['edges']]:
+    if m is None:
+      continue
+    for match in itertools.chain(re.finditer(g3ReviewerRegex, m),
+                                 re.finditer(beamReviewerRegex, m)):
+      for user in re.finditer(usernameRegex, match.groups()[0]):
+        # First group decides if it is additive or subtractive
+        remove = user.groups()[0] == '-'
```

Comment:
This is interested; I haven't heard of people using a convention to remove reviewers, but it seems nice.

Reply:
I've seen it couple of times. Although, I'm not sure it will be reflected properly on history as-is, so I do not utilize this in alternative reviewers calculation logic. We can build more on-top of it later.